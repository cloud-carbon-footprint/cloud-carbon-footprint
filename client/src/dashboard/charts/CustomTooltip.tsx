/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

/*
We're using a custom tool tip so that we can show data from the wattHours and cost series in the tooltip,
even when the wattHours and cost series are toggled off. When using the default tooltip provided by apex charts,
there is no way to access additional data in the series, for example, the wattHours and cost data in the co2e series.
The custom tooltip also allows us to add an asterisk for data points that were estimated using an average CPU constant.
 */

import React, { ReactElement } from 'react'
import moment from 'moment'
import { cloudEstPerDay } from '../../models/types'

export const CustomTooltip = ({
  data,
  dataPointIndex,
}: {
  data: cloudEstPerDay[]
  dataPointIndex: number
}): ReactElement => {
  const date = moment.utc(data[dataPointIndex].x).format('MMMM DD')
  return (
    <div style={{ padding: '10px' }}>
      <div>
        <b>{date}</b>
      </div>
      {data[dataPointIndex].usesAverageCPUConstant ? (
        <div>{data[dataPointIndex].y} mt CO2e*</div>
      ) : (
        <div> {data[dataPointIndex].y} mt CO2e</div>
      )}
      <div>{data[dataPointIndex].wattHours} watt hrs</div>
      <div>${data[dataPointIndex].cost} cost</div>
    </div>
  )
}
