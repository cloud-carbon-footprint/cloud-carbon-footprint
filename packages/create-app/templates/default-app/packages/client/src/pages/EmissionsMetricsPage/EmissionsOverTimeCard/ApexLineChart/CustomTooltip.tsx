/*
 * Â© 2021 Thoughtworks, Inc.
 */

/*
We're using a custom tool tip so that we can show data from the wattHours and cost series in the tooltip,
even when the wattHours and cost series are toggled off. When using the default tooltip provided by apex charts,
there is no way to access additional data in the series, for example, the wattHours and cost data in the co2e series.
The custom tooltip also allows us to add an asterisk for data points that were estimated using an average CPU constant.
 */

import React, { ReactElement } from 'react'
import moment from 'moment'
import { cloudEstPerDay } from '../../../../Types'

type CustomTooltipProps = {
  dataPoint: cloudEstPerDay
  grouping: string
}

const CustomTooltip = ({
  dataPoint,
  grouping,
}: CustomTooltipProps): ReactElement => {
  if (dataPoint?.x) {
    const dateLabel = {
      day: moment.utc(dataPoint.x).format('MMM DD, YYYY'),
      week: moment.utc(dataPoint.x).format('[Week] w, MMM'),
      month: moment.utc(dataPoint.x).format('MMM YYYY'),
      quarter: moment.utc(dataPoint.x).format('Qo [Quarter] YYYY'),
      year: moment.utc(dataPoint.x).format('YYYY'),
    }
    return (
      <div style={{ padding: '10px' }}>
        <div>
          <b>{dateLabel[grouping]}</b>
        </div>
        <div>
          {dataPoint.y} metric tons CO2e
          {dataPoint.usesAverageCPUConstant && '*'}
        </div>
        <div>{dataPoint.kilowattHours} kilowatt hrs</div>
        <div>${dataPoint.cost} cost</div>
      </div>
    )
  }
  return <div />
}

export default CustomTooltip
