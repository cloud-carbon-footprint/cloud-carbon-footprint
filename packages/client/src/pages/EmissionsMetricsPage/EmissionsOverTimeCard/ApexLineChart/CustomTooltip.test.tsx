/*
 * Â© 2021 Thoughtworks, Inc.
 */

import each from 'jest-each'
import { create } from 'react-test-renderer'
import { render } from '@testing-library/react'
import CustomTooltip from './CustomTooltip'

describe('Custom Tooltip', () => {
  const cloudEstimatesPerDay = [
    {
      x: new Date('2020-12-01'),
      y: 10,
      usesAverageCPUConstant: false,
      kilowattHours: 100,
      cost: 5,
    },
    {
      x: new Date('2020-12-02'),
      y: 20,
      usesAverageCPUConstant: true,
      kilowattHours: 200,
      cost: 10,
    },
    {
      x: new Date('2021-11-01'),
      y: 20,
      usesAverageCPUConstant: true,
      kilowattHours: 200,
      cost: 10,
    },
  ]

  it('renders data in tooltip', () => {
    const tooltip = create(
      <CustomTooltip dataPoint={cloudEstimatesPerDay[1]} grouping="day" />,
    )
    expect(tooltip.toJSON()).toMatchSnapshot()
  })

  it('should return empty tooltip if no data point', () => {
    const { container } = render(
      <CustomTooltip dataPoint={undefined} grouping="day" />,
    )
    expect(container.firstChild).toBeEmpty()
  })

  it('should show an asterisk in label for a data point that uses an average CPU constant', () => {
    const { getByText } = render(
      <CustomTooltip dataPoint={cloudEstimatesPerDay[1]} grouping="day" />,
    )
    const expectedLabel = `${cloudEstimatesPerDay[1].y} metric tons CO2e*`

    expect(getByText(expectedLabel)).toBeInTheDocument()
  })

  it('should not show an asterisk for a data point that do not use an average CPU constant', () => {
    const { getByText } = render(
      <CustomTooltip dataPoint={cloudEstimatesPerDay[0]} grouping="day" />,
    )
    const expectedLabel = `${cloudEstimatesPerDay[0].y} metric tons CO2e`

    expect(getByText(expectedLabel)).toBeInTheDocument()
  })

  describe('Show tooltip title according to group by param', () => {
    each([
      ['2021', 'year'],
      ['4th Quarter 2021', 'quarter'],
      ['Nov 2021', 'month'],
      ['Week 45, Nov', 'week'],
      ['Nov 01, 2021', 'day'],
    ]).it('should show %s when grouping by %s', (expectedLabel, grouping) => {
      const { getByText } = render(
        <CustomTooltip
          dataPoint={cloudEstimatesPerDay[2]}
          grouping={grouping}
        />,
      )

      expect(getByText(expectedLabel)).toBeInTheDocument()
    })
  })
})
