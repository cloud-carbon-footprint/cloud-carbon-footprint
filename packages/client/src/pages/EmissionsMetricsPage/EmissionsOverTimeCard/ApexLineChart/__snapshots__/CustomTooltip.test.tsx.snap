// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Tooltip renders data in tooltip 1`] = `
<div
  style={
    Object {
      "padding": "10px",
    }
  }
>
  <div>
    <b>
      Dec 02, 2020
    </b>
  </div>
  <div>
    20
     metric tons CO2e
    *
  </div>
  <div>
    200
     kilowatt hrs
  </div>
  <div>
    $
    10
     cost
  </div>
</div>
`;
